### ISSUE TITLE
Inclusion Functions inside Shard Committee Code Block

### ISSUE THREAD

**Issue by terenc3t** [ethereum/eth2.0-specs#396]
During per-epoch processing, when we first loop through `shard_committee`, I understand we want to track each committee's total attesting balance and total balance. What's the purpose of `inclusion_slot` and `inclusion_distance` in the end?

```

For every shard_committee in state.shard_committees_at_slots:
   Let attesting_validator_indices(shard_committee, shard_block_root) be the union of the validator index sets given by [get_attestation_participants(state, a.data, a.participation_bitfield) for a in current_epoch_attestations + previous_epoch_attestations if a.shard == shard_committee.shard and a.shard_block_root == shard_block_root].
   ...
   Let inclusion_slot(state, index) = a.slot_included for the attestation a where index is in get_attestation_participants(state, a.data, a.participation_bitfield).
   Let inclusion_distance(state, index) = a.slot_included - a.data.slot where a is the above attestation.
```

**Comment by ralexstokes** [comment]
There are two things going on here:

1. Each validator gets some reward that is proportional to how quickly they can get their attestations into the chain (the `inclusion_distance`). You can see this directly being used to calculate the "inclusion distance" reward type under [Justification and finalization](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#justification-and-finalization). The "sense" of this reward (i.e. is it a gain or is it a loss?) depends on how long it has been since we last finalized an epoch. If it has only been a few epochs since finality, then you gain a reward that decays linearly w.r.t the number of slots and in the limit goes to 0. If it has been many epochs, then you gain a loss that increases as the  `inclusion_distance` increases, tending towards losing the full `base_reward`.

The purpose of this mechanism is to encourage validators to make attestations in as timely a manner as possible. Attestation generation is probably the most critical piece in terms of network health -- things start stalling if we don't see a critical mass of attestations for some time.

2. for every relevant slot, the block proposer for that slot gains some small reward for including the attestations. you determine which proposer is eligible for the reward based on the `inclusion slot`.

The purpose of this mechanism is to encourage block proposers to actually include attestations in the chain. They could simply omit them or only include a subset (at the expense of an "honest minority" in the case of an attack, for example). The idea w/ this reward is that the price of the reward dominates the gain from any of these censorship scenarios -- again to ensure a healthy network.

**Comment by terenc3t** [comment]
Clearly explained. Thank you! I'm still unclear how `inclusion_slot` and `inclusion_distance` connect to `For every shard_committee_at_slot in state.shard_committees_at_slots and for every shard_committee in shard_committee_at_slot`

`Let inclusion_slot(state, index) = a.slot_included for the attestation a where index is in get_attestation_participants(state, a.data, a.participation_bitfield)`

What is the attestation in the shard_committee code block?

**Comment by ralexstokes** [comment]
their usage is unrelated to the crosslinking logic so i'm just reading them as helpers that happen to be there. i agree the indentation/bulleting is confusing and we should probably update it.

here is what is going on, note this is just pseudocode, i.e. there are prolly opportunities for caching various reads of the state to get a more performant implementation...

```python
## Helpers

## refer to spec for `get_attestation_participants`, `get_beacon_proposer_index`, `base_reward` and constants

def inclusion_slot(a: PendingAttestationRecord) -> uint:
    return a.slot_included

def inclusion_distance(a: PendingAttestationRecord) -> int:
    return a.slot_included - a.data.slot


## Processing routines per epoch

# update crosslinks based on attestations made for this epoch
# dispense rewards for new crosslinks made

## independently of the crosslinks, do the following based on attestations made:

def process_attestation_inclusion_rewards(state: BeaconState):
    previous_epoch_attestations = [a for a in state.latest_attestations if state.slot - 2 * EPOCH_LENGTH <= a.slot < state.slot - EPOCH_LENGTH]
    for a in previous_epoch_attestations:
        proposer_index = get_beacon_proposer_index(state, inclusion_slot(a))
        for index in get_attestation_participants(state, a.data, a.participation_bitfield):
            state.validator_balances[proposer_index] += base_reward(state, index) // INCLUDER_REWARD_QUOTIENT


#  run only one of the following two routines depending on how many epochs it has been since last finality; refer to spec for more info

def process_inclusion_distance_reward(state: BeaconState):
    previous_epoch_attestations = [a for a in state.latest_attestations if state.slot - 2 * EPOCH_LENGTH <= a.slot < state.slot - EPOCH_LENGTH]
    for a in previous_epoch_attestations:
        for index in get_attestation_participants(state, a.data, a.participation_bitfield):
            state.validator_balances[index] += base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY // inclusion_distance(a)

def process_inclusion_distance_penalty(state: BeaconState):
    previous_epoch_attestations = [a for a in state.latest_attestations if state.slot - 2 * EPOCH_LENGTH <= a.slot < state.slot - EPOCH_LENGTH]
    for a in previous_epoch_attestations:
        for index in get_attestation_participants(state, a.data, a.participation_bitfield):
            state.validator_balances[index] -= (base_reward(state, index) - base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY // inclusion_distance(a))
```

if i'm misreading or made a mistake someone plz correct me :)

**Comment by terenc3t** [comment]
The pseudocode is the same as what I have in mind. Thank you for drawing it out.
I'm glad we both agree on `inclusion_slot` and `inclusion_distance` are incorrectly indented. I'll open a PR to fix it = )

**Comment by djrtwo** [comment]
closed via #406
